pkg/provider-local/controller/operatingsystemconfig/actuator.go

```diff
@@ -0,0 +1,124 @@
+	//go:embed templates/cloud-init.template
+	cloudInitTemplateString string
+	cloudInitGenerator      *ostemplate.CloudInitGenerator
+)
+
+func init() {
+	cloudInitTemplate, err := ostemplate.NewTemplate("cloud-init").Parse(cloudInitTemplateString)
+	runtime.Must(err)
+
+	cloudInitGenerator = ostemplate.NewCloudInitGenerator(
+		cloudInitTemplate,
+		ostemplate.DefaultUnitsPath,
+		"/usr/bin/env bash %s",
+		func(*extensionsv1alpha1.OperatingSystemConfig) (map[string]interface{}, error) {
+			return nil, nil
+		},
+	)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) ([]byte, *string, []string, []string, []extensionsv1alpha1.Unit, []extensionsv1alpha1.File, error) {
+	if !a.useGardenerNodeAgent {
+		cloudConfig, cmd, err := oscommonactuator.CloudConfigFromOperatingSystemConfig(ctx, log, a.client, osc, cloudInitGenerator)
+		if err != nil {
+			return nil, nil, nil, nil, nil, nil, fmt.Errorf("could not generate cloud config: %w", err)
+		}
+		return cloudConfig, cmd, oscommonactuator.OperatingSystemConfigUnitNames(osc), oscommonactuator.OperatingSystemConfigFilePaths(osc), nil, nil, nil
+	}
+
+	switch purpose := osc.Spec.Purpose; purpose {
+	case extensionsv1alpha1.OperatingSystemConfigPurposeProvision:
+		userData, err := a.handleProvisionOSC(ctx, osc)
+		return []byte(userData), nil, nil, nil, nil, nil, err
+
+	case extensionsv1alpha1.OperatingSystemConfigPurposeReconcile:
+		additionalUnits, additionalFiles, err := a.handleReconcileOSC(osc)
+		return nil, nil, nil, nil, additionalUnits, additionalFiles, err
+
+	default:
+		return nil, nil, nil, nil, nil, nil, fmt.Errorf("unknown purpose: %s", purpose)
+	}
+}
+
+func (a *actuator) Delete(_ context.Context, _ logr.Logger, _ *extensionsv1alpha1.OperatingSystemConfig) error {
+	return nil
+}
+
+func (a *actuator) Migrate(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) error {
+	return a.Delete(ctx, log, osc)
+}
+
+func (a *actuator) ForceDelete(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) error {
+	return a.Delete(ctx, log, osc)
+}
+
+func (a *actuator) Restore(ctx context.Context, log logr.Logger, osc *extensionsv1alpha1.OperatingSystemConfig) ([]byte, *string, []string, []string, []extensionsv1alpha1.Unit, []extensionsv1alpha1.File, error) {
+	return a.Reconcile(ctx, log, osc)
+}
+
+func (a *actuator) handleProvisionOSC(ctx context.Context, osc *extensionsv1alpha1.OperatingSystemConfig) (string, error) {
+	writeFilesToDiskScript, err := operatingsystemconfig.FilesToDiskScript(ctx, a.client, osc.Namespace, osc.Spec.Files)
+	if err != nil {
+		return "", err
+	}
+	writeUnitsToDiskScript := operatingsystemconfig.UnitsToDiskScript(osc.Spec.Units)
+
+	return `#!/bin/bash
+` + writeFilesToDiskScript + `
+` + writeUnitsToDiskScript + `
+systemctl daemon-reload
+systemctl enable gardener-node-agent && systemctl restart gardener-node-agent
```

Comment:
```suggestion
systemctl enable gardener-node-init && systemctl start gardener-node-init
```
Isn't it `gardener-node-init` systemd unit which should be started to bootstrap gardener-node-agent and which enables its service?

Reply:
Nice catch <3